content/development/contribution_guidelines.md

```diff
@@ -109,17 +109,107 @@ The older style of representing a backslash using the character class `[\\\\]` m
+
+Using lazy quantifiers, the regular expression engine will instead match _as few characters as possible_. The engine will match ` `, then look for `;` and will not find it. The match will then be expanded to ` =` and, again, a match of `;` is attempted. This continues until the match is ` = "name=evil` and the engine finds `;`. While lazy matching still includes some work, in this case, backtracking would require many more steps.
+
+Lazy matching can have the inverse effect though. Consider the following expression:
+```python
+"@rx (?i)\b(?:s(?:tyle|rc)|href)\b[\s\S]*?="
+```
+It matches some HTML attributes and then expects to see `=`. Using a somewhat contrived input, the lazy qantifier will require more steps to match then the greedy version would: `style                     =`. With the lazy quantifier, the regular expression engine will expand the match by one character for each of the space characters in the input, that are 21 steps in this case. With the greedy quantifier, the engine would match up to the end in a single step, backtrack one character and then match `=` (note that `=` is included in `[\s\S]`), which makes 3 steps.
+
+To summarize: be very mindful about when and why you use lazy quantifiers in your regular expressions.
+
+### Writing Regular Expressions for Non-backtracking Compatibility
+
+Traditional regualar expression engines use backtracking to solve some additional problems, such as finding a string that is preceded or followed by another string. While this functionality can certainly come in handy and has it's place in certain applications, it can also lead to performance issues and, in uncontrolled environments, open up possibilities for attacks (the term "ReDoS" is often used to describe an attack that exhausts process or system resources due to excessive backtracking).
 
-### Writing RE2-compatible Regular Expressions
+The OWASP Core Rule Set tries to be compatible with non-backtracking regular expression engines, such as re2 because
+- non-backtracking engines are less vulnerable to ReDoS attacks
+- non-backtracking engines can often outperform backtracking engines
+- we want to leave the choice of the engine to the user / system.
 
-Avoid "lookbehind" and "lookafter".
+To ensure compatibility with non-backtracking regular expression engines, the following operations are not permitted in regular expressions:
+- positive lookahead (e.g.  `(?=regex)`)
+- negative lookahead (e.g. `(?!regex)`)
+- positive lookbehind (e.g. `(?<=regex)`)
+- negative lookbehind (e.g. `(?<!regex)`)
+- named capture groups (e.g. `(?P<name>regex)`)
+- backreferences (e.g. `\1`)
+- named backreferences (e.g. `(?P=name)`)
+- conditionals (e.g. `(?(regex)then|else)`)
+- recursive calls to capture groups (e.g. `(?1)`)
 
-TODO
+This list is not exhaustive but covers the most important points.
```

Comment:
What about adding a list to upstream like https://github.com/google/re2/wiki/Syntax ?

Reply:
Good idea.